### PR TITLE
[FLOC-1595] Partial state updates

### PR DESCRIPTION
**Pull request by itamarst**
https://clusterhq.atlassian.net/browse/FLOC-1595

### CODE REVIEW

File: flocker/control/_clusterstate.py [hunk 1]
```diff
@@ -45,7 +41,9 @@ def manifestation_path(self, hostname, dataset_id):
 
         :return FilePath: The path where the manifestation exists.
         """
-        return self._nodes[hostname].paths[dataset_id]
+        node, = (node for node in self._deployment_state.nodes
```
Comment (exarkun) [inline flocker/control/_clusterstate.py]: `[node]` reads less like dirt on my screen than `node,`

File: flocker/control/_model.py [hunk 2]
```diff
@@ -517,6 +517,32 @@ class DeploymentState(PRecord):
...
+
+        :param NodeState node: An update for ``NodeState`` with same
+             hostname in this ``DeploymentState``.
+
+        :return DeploymentState: Updated with new ``NodeState``.
+        """
+        nodes = {n for n in self.nodes if n.hostname == node_state.hostname}
+        if not nodes:
+            return self.transform(["nodes"], lambda s: s.add(node_state))
+        updated_node, = nodes
```
Comment (exarkun) [inline flocker/control/_model.py]: `[...]` pls

File: flocker/control/_model.py [hunk 3]
```diff
@@ -517,6 +517,32 @@ class DeploymentState(PRecord):
...
+        for key, value in node_state.items():
+            if value is not None:
+                updated_node = updated_node.set(key, value)
+        return DeploymentState(nodes=frozenset(
```
Comment (exarkun) [inline flocker/control/_model.py]: Doesn't need to be a `frozenset` anymore, right?

File: flocker/control/_model.py [hunk 4]
```diff
@@ -517,6 +517,32 @@ class DeploymentState(PRecord):
...
+            list(n for n in self.nodes if n.hostname != node_state.hostname) +
+            [updated_node]))
```
Comment (exarkun) [inline flocker/control/_model.py]: How about `return self.set(nodes=nodes.discard(original_node).add(updated_node))`?

File: flocker/control/test/test_model.py [hunk 5]
```diff
@@ -702,3 +720,77 @@ class Record(PRecord):
...
+            update_manifestations)
+        self.assertEqual(updated,
+                         DeploymentState(nodes=[NodeState(
+                             hostname=u"node1.example.com",
+                             applications=frozenset({Application(
+                                 name=u'site-clusterhq.com',
+                                 image=DockerImage.from_string(u"image"))}),
+                             used_ports=[1, 2],
+                             paths={dataset_id: FilePath(b"/xxx")},
+                             manifestations={dataset_id: manifestation})]))
```
Comment (exarkun) [inline flocker/control/test/test_model.py]: Some duplication could be avoided by defining this expected `DeploymentState` first and then defining `update_applications` and `update_manifestations` in terms of it (eg `update_applications = expected.set(manifestations=None, paths=None)`.

File: flocker/control/_clusterstate.py [hunk 6]
```diff
@@ -17,13 +17,9 @@ class ClusterStateService(Service):
...
     semantics of expiring data, which should happen so stale information
     isn't treated as correct.
-
-    https://clusterhq.atlassian.net/browse/FLOC-1542 will deal with
-    NodeState that has manifestations or application set to ``None``; for
-    now we assume all data is present in any given update.
     """
     def __init__(self):
-        self._nodes = {}
+        self._deployment_state = DeploymentState()
```
Comment (exarkun) [inline flocker/control/_clusterstate.py]: ivar documentation for this attribute missing

**Comment by exarkun** [comment]
Thanks.  This looks good.  One thing I couldn't help thinking about while reading this is that we should make `DeploymentState.nodes` a pmap instead of a pset.

Address inline comments to your satisfaction and then merge.

**Comment by itamarst** [comment]
Yeah, the `pmap` thing occurred to me too... but then it's more different than `Deployment`. But maybe that's OK? Maybe in some future issue, although possibly at this point I'm done missing with it for this sprint.

File: flocker/control/_model.py (hunk 4 above)
Comment (itamarst) [inline flocker/control/_model.py]: HOORJ

File: flocker/control/_clusterstate.py (hunk 1 above)
Comment (itamarst) [inline flocker/control/_clusterstate.py]: "Ctrl-+"?

**Comment by exarkun** [comment]
Later is fine, yea.  And maybe we can change `Deployment` in the same way at the same time.

**Pull request merged**